### PR TITLE
Seek for messages in topics

### PR DESCRIPTION
**Pull request by albertlast**
This is the next try to implement seek for pages in topics.
Handle four cases:
next page
before page
current page (when someone push refresh)
start page (it's mostly the same as current page)

@Sesquipedalian can you please look for issue of this version?

### CODE REVIEW

**Comment by Sesquipedalian** [comment]
Sorry, I got swamped with real life stuff to take care of. I will take a look at this as soon as I can. :)

**Comment by albertlast** [comment]
When it take to long we could merge this pr and
let everyone test this.
in the worst case do a revert again.

File: Sources/Display.php [hunk 1]
```diff
@@ -842,45 +842,163 @@ function Display()
...
+		$start_char = 'C';
+		$page_id = $_SESSION['page_first_id'];
+	}
+	elseif ($start == 0) //special case start page
+	{	// special case start page
+		$start_char = 'C';
+		$page_id = $context['topicinfo']['id_first_msg'];
+	}
+	else
+		$start_char = null;
```
Comment (Sesquipedalian) [inline Sources/Display.php]: This would avoid checking the value of $_SESSION['page_topic'] more than once:
```
	// Check if we can use the seek method to speed things up
	if (isset($_SESSION['page_topic']) && $_SESSION['page_topic'] == $topic)
	{
		// User moved to the next page
		if (isset($_SESSION['page_next_start']) && $_SESSION['page_next_start'] == $start)
		{	
			$start_char = 'M'; 
			$page_id = $_SESSION['page_last_id'];
		}
		// User moved to the previous page
		elseif (isset($_SESSION['page_before_start']) && $_SESSION['page_before_start'] == $start)
		{	
			$start_char = 'L';
			$page_id = $_SESSION['page_first_id'];
		}
		// User refreshed the current page
		elseif (isset($_SESSION['page_current_start']) && $_SESSION['page_current_start'] == $start)
		{
			$start_char = 'C';
			$page_id = $_SESSION['page_first_id'];
		}
	}
	// Special case start page
	elseif ($start == 0)
	{
		$start_char = 'C';
		$page_id = $context['topicinfo']['id_first_msg'];
	}
	else
		$start_char = null;
```
Also, this would make the comments a bit clearer in English and would format them according to SMF's standard coding style. :)

File: Sources/Display.php [hunk 2]
```diff
@@ -842,45 +842,163 @@ function Display()
...
+		if ($smcFunc['db_num_rows']($request) < 1)
+			unset($start_char);
+		else
+		{
+			while ($row = $smcFunc['db_fetch_assoc']($request))
+			{
+				if ($row['id_msg'] != $page_id || $start_char === 'C')
+				{
+					if (!empty($row['id_member']))
+						$all_posters[$row['id_msg']] = $row['id_member'];
```
Comment (Sesquipedalian) [inline Sources/Display.php]: Nit: please add a line break after this to improve legibility.

File: Sources/Display.php [hunk 3]
```diff
@@ -842,45 +842,163 @@ function Display()
...
+					$messages[] = $row['id_msg'];
+
+					if ($start_char === 'C' && $row['id_msg'] == $page_id) // find the start message of current page
+						$found_msg = true;
+				}
+				else
+					$found_msg = true; // find the start message of next or before page
```
Comment (Sesquipedalian) [inline Sources/Display.php]: This else statement is probably a bug. It will set $found_msg to true in a number of cases where I doubt that is what you really intended. For example, if the first retrieved row from the database matched the previous if statement, then the rest of the retrieved rows probably won't match. The result is that $found_msg is probably *always* going to be true by the time you finish looping through all the retrieved rows.

File: Sources/Display.php [hunk 4]
```diff
@@ -842,45 +842,163 @@ function Display()
...
+		);
+
+		while ($row = $smcFunc['db_fetch_assoc']($request))
+		{
+			if (!empty($row['id_member']))
+				$all_posters[$row['id_msg']] = $row['id_member'];
+			$messages[] = $row['id_msg'];
+		}
+
+		// get the message in this order how they shown
```
Comment (Sesquipedalian) [inline Sources/Display.php]: Suggested: `// Sort the messages into the correct display order`

File: Sources/Display.php [hunk 5]
```diff
@@ -842,45 +842,163 @@ function Display()
...
+		if (!$ascending)
+			sort($messages);
 	}
+
+	// Save the next start of the next page and the end of the page before
```
Comment (Sesquipedalian) [inline Sources/Display.php]: Suggested: `// Remember the paging data for next time`

**Review by Sesquipedalian — CHANGES_REQUESTED** [review]
Looks pretty great. I've made a few suggestions, and I found one spot where the logic to deal with unexpected results appears to be a bit buggy. I could not find a way to turn that into a problematic result, but you should still probably take a look at it.

File: Sources/Display.php (hunk 3 above)
Comment (albertlast) [reply Sources/Display.php]: The else cover this case:
you go the next or previous page and
the page_id was deleted/get lost some how.

Testcase: go to a thread with 3 pages
go to page 3, delete somehow the first msg on page three and call then page 2 than this
fallback get active.

To make it clear i describe now the same thing from coding side:
Common is that we need the amount of msg_id which is needed to show ->
when you want to show 15 msg = limit 15.
To cover this case described above we set limit + 1,
so the sql return 15 + (start page_id) we need this for check but we don't want
this deliver to the next logic part.
In short we don't want the start page_id in $messages.

**Comment by albertlast** [comment]
Some how i didn't see all you comment at once,
so i created multiple commits for you comments.

I hope i cover all the things.

File: Sources/Display.php (hunk 3 above)
Comment (Sesquipedalian) [reply Sources/Display.php]: Okay, I think I see what you mean. In that case, you can probably simplify your code a bit by replacing this:
```
if ($row['id_msg'] != $page_id || $start_char === 'C')
{
	if (!empty($row['id_member']))
		$all_posters[$row['id_msg']] = $row['id_member'];

	$messages[] = $row['id_msg'];

	if ($start_char === 'C' && $row['id_msg'] == $page_id) // find the start message of current page
		$found_msg = true;
}
else
	$found_msg = true; // find the start message of next or before page
```
with this:
```
if ($row['id_msg'] == $page_id)
	$found_msg = true;

if ($start_char === 'C' || !$found_msg)
{
	if (!empty($row['id_member']))
		$all_posters[$row['id_msg']] = $row['id_member'];

	$messages[] = $row['id_msg'];
}
```

They are logically equivalent, but the second version is easier to understand.

File: Sources/Display.php [hunk 6]
```diff
@@ -842,45 +842,171 @@ function Display()
...
+	
 	$messages = array();
 	$all_posters = array();
-	while ($row = $smcFunc['db_fetch_assoc']($request))
+	
+	if (isset($start_char))
+	{
+		$firstIndex = 0;
+
+		if ($start_char === 'M' or $start_char === 'S' or $start_char === 'C')
```
Comment (Sesquipedalian) [inline Sources/Display.php]: Should `$start_char === 'S'` be `$start_char === 'L'` here?

File: Sources/Display.php (hunk 3 above)
Comment (albertlast) [reply Sources/Display.php]: When I read correctly (for me is this version harder to  read)
```
if ($row['id_msg'] == $page_id)
	$found_msg = true;

if ($start_char === 'C' || !$found_msg)
{
	if (!empty($row['id_member']))
		$all_posters[$row['id_msg']] = $row['id_member'];

	$messages[] = $row['id_msg'];
}
```
Doesn't skip the 16th dataset for the L and M case.
```
if ($row['id_msg'] == $page_id)
	$found_msg = true;

if ($start_char === 'C' || $row['id_msg'] != $page_id)
{
	if (!empty($row['id_member']))
		$all_posters[$row['id_msg']] = $row['id_member'];

	$messages[] = $row['id_msg'];
}
```
This would be logically equivalent in my eyes.

Comment (Sesquipedalian) [reply Sources/Display.php]: Fair enough. :)

**Comment by albertlast** [comment]
So can we merge this?

**Review by Sesquipedalian — APPROVED** [review]

**Comment by Sesquipedalian** [comment]
Nice to see this go in. :)
